diff -ur rsync-2.6.9/batch.c rsync-2.6.9-progress/batch.c
--- rsync-2.6.9/batch.c	2006-04-26 01:51:12.000000000 +0200
+++ rsync-2.6.9-progress/batch.c	2010-03-04 15:57:39.000000000 +0100
@@ -173,6 +173,8 @@
 	int fd, i, len;
 	char *p, filename[MAXPATHLEN];
 
+	if (strcmp(batch_name, "/dev/null") == 0) return;
+
 	stringjoin(filename, sizeof filename,
 		   batch_name, ".sh", NULL);
 	fd = do_open(filename, O_WRONLY | O_CREAT | O_TRUNC,
diff -ur rsync-2.6.9/io.c rsync-2.6.9-progress/io.c
--- rsync-2.6.9/io.c	2006-10-01 00:11:20.000000000 +0200
+++ rsync-2.6.9-progress/io.c	2010-03-04 15:59:54.000000000 +0100
@@ -51,6 +51,7 @@
 extern char *filesfrom_host;
 extern struct stats stats;
 extern struct file_list *the_file_list;
+extern int batch_dev_null;
 
 const char phase_unknown[] = "unknown";
 int ignore_timeout = 0;
@@ -1079,7 +1080,7 @@
 			continue;
 		}
 
-		if (FD_ISSET(fd, &e_fds)) {
+		if (FD_ISSET(fd, &e_fds) && !(fd == batch_fd && batch_dev_null)) {
 			rsyserr(FINFO, errno,
 				"select exception on fd %d", fd);
 		}
diff -ur rsync-2.6.9/main.c rsync-2.6.9-progress/main.c
--- rsync-2.6.9/main.c	2006-10-14 01:46:32.000000000 +0200
+++ rsync-2.6.9-progress/main.c	2010-03-04 16:35:00.000000000 +0100
@@ -71,6 +71,7 @@
 extern char curr_dir[MAXPATHLEN];
 extern struct filter_list_struct server_filter_list;
 
+int batch_dev_null = 0;
 int local_server = 0;
 int new_root_dir = 0;
 mode_t orig_umask = 0;
@@ -1347,6 +1348,7 @@
 		if (read_batch && strcmp(batch_name, "-") == 0)
 			batch_fd = STDIN_FILENO;
 		else {
+			if (strcmp(batch_name, "/dev/null") == 0) batch_dev_null = 1;
 			batch_fd = do_open(batch_name,
 				   write_batch ? O_WRONLY | O_CREAT | O_TRUNC
 				   : O_RDONLY, S_IRUSR | S_IWUSR);
diff -ur rsync-2.6.9/progress.c rsync-2.6.9-progress/progress.c
--- rsync-2.6.9/progress.c	2006-10-14 01:17:33.000000000 +0200
+++ rsync-2.6.9-progress/progress.c	2010-03-04 15:57:39.000000000 +0100
@@ -105,9 +105,9 @@
 			stats.num_files);
 	} else
 		strlcpy(eol, "\r", sizeof eol);
-	rprintf(FCLIENT, "%12s %3d%% %7.2f%s %4d:%02d:%02d%s",
+	rprintf(FCLIENT, "%12s %3d%% %7.2f%s %4d:%02d:%02d %lli %lli%s",
 		human_num(ofs), pct, rate, units,
-		remain_h, remain_m, remain_s, eol);
+		remain_h, remain_m, remain_s, stats.total_read, stats.total_written, eol);
 }
 
 void end_progress(OFF_T size)
